### PR TITLE
Fix bug when data/values ownership is incorrectly assigned.

### LINKED ISSUE #123: "update" functions assign data ownership incorrectly
Linear algebra methods that update data in local objects check if memory is allocated and if not they allocate memory underneath so they can copy/update data on user-specified device. However, they assign ownership of that data upon copy and not upon allocation. If the data exists and is not owned by the local object, the ownership will be incorrectly assigned to the local who will eventually try to delete it. This could cause double-delete segfault.

### PR DESCRIPTION
**Pull request by pelesh**
Closes #123

### CODE REVIEW

File: resolve/vector/VectorHandlerCuda.cpp [hunk 1]
```diff
@@ -49,8 +49,10 @@ namespace ReSolve {
   { 
     cublasHandle_t handle_cublas =  workspace_->getCublasHandle();
     double nrm = 0.0;
-    cublasStatus_t st= cublasDdot (handle_cublas,  x->getSize(), x->getData(memory::DEVICE), 1, y->getData(memory::DEVICE), 1, &nrm);
-    if (st!=0) {printf("dot product crashed with code %d \n", st);}
+    cublasStatus_t st= cublasDdot(handle_cublas,  x->getSize(), x->getData(memory::DEVICE), 1, y->getData(memory::DEVICE), 1, &nrm);
```
Comment (kswirydo) [inline resolve/vector/VectorHandlerCuda.cpp:52]:
```suggestion
    cublasStatus_t st = cublasDdot(handle_cublas,  x->getSize(), x->getData(memory::DEVICE), 1, y->getData(memory::DEVICE), 1, &nrm);
```

File: resolve/vector/VectorHandlerCuda.cpp [hunk 2]
```diff
@@ -49,8 +49,10 @@ namespace ReSolve {
...
+    if (st!=0) {
```
Comment (kswirydo) [inline resolve/vector/VectorHandlerCuda.cpp:53]:
```suggestion
    if (st != 0) {
```

File: resolve/vector/VectorHandlerCuda.cpp [hunk 3]
```diff
@@ -66,7 +68,7 @@ namespace ReSolve {
     cublasHandle_t handle_cublas =  workspace_->getCublasHandle();
     cublasStatus_t st = cublasDscal(handle_cublas, x->getSize(), alpha, x->getData(memory::DEVICE), 1);
     if (st!=0) {
```
Comment (kswirydo) [inline resolve/vector/VectorHandlerCuda.cpp:70]:
```suggestion
    if (st! = 0) {
```

File: tests/functionality/testKLU_GLU.cpp [hunk 4]
```diff
@@ -146,9 +146,9 @@ int main(int argc, char *argv[])
   status = matrix_handler->matvec(A, vec_x, vec_r, &ONE, &MINUSONE,"csr", ReSolve::memory::HOST);
   error_sum += status;
  
-  real_type normRmatrix1CPU = sqrt(vector_handler->dot(vec_r, vec_r, ReSolve::memory::DEVICE));
+  real_type normRmatrix1CPU = sqrt(vector_handler->dot(vec_r, vec_r, ReSolve::memory::HOST));
  
-  std::cout<<"Results (first matrix): "<<std::endl<<std::endl;
+  std::cout<<"Results (first matrix): "<<std::endl<<std::endl << std::scientific;
```
Comment (kswirydo) [inline tests/functionality/testKLU_GLU.cpp:151]:
```suggestion
  std::cout<<"Results (first matrix): " << std::endl << std::endl << std::scientific;
```

File: tests/functionality/testKLU_GLU.cpp [hunk 5]
```diff
@@ -146,9 +146,9 @@ int main(int argc, char *argv[])
...
   std::cout<<"\t ||b-A*x||_2                 : " << std::setprecision(16) << normRmatrix1    << " (residual norm)" << std::endl;
```
Comment (kswirydo) [inline tests/functionality/testKLU_GLU.cpp:152]:
```suggestion
  std::cout << "\t ||b-A*x||_2                 : " << std::setprecision(16) << normRmatrix1    << " (residual norm)" << std::endl;
```

File: tests/functionality/testKLU_GLU.cpp [hunk 6]
```diff
@@ -146,9 +146,9 @@ int main(int argc, char *argv[])
...
   std::cout<<"\t ||b-A*x||_2  (CPU)          : " << std::setprecision(16) << normRmatrix1CPU << " (residual norm)" << std::endl;
```
Comment (kswirydo) [inline tests/functionality/testKLU_GLU.cpp:153]:
```suggestion
  std::cout << "\t ||b-A*x||_2  (CPU)          : " << std::setprecision(16) << normRmatrix1CPU << " (residual norm)" << std::endl;
```

File: tests/functionality/testKLU_GLU.cpp [hunk 7]
```diff
@@ -146,9 +146,9 @@ int main(int argc, char *argv[])
...
   std::cout<<"\t ||b-A*x||_2/||b||_2         : " << normRmatrix1/normB1   << " (scaled residual norm)"             << std::endl;
```
Comment (kswirydo) [inline tests/functionality/testKLU_GLU.cpp:154]:
```suggestion
  std::cout << "\t ||b-A*x||_2/||b||_2         : " << normRmatrix1/normB1   << " (scaled residual norm)"             << std::endl;
```

File: tests/functionality/testKLU_RocSolver_FGMRES.cpp [hunk 8]
```diff
@@ -141,7 +141,7 @@ int main(int argc, char *argv[])
 
   real_type normRmatrix1CPU = sqrt(vector_handler->dot(vec_r, vec_r, ReSolve::memory::DEVICE));
 
-  std::cout<<"Results (first matrix): "<<std::endl<<std::endl;
+  std::cout<<"Results (first matrix): " << std::scientific <<std::endl<<std::endl;
```
Comment (kswirydo) [inline tests/functionality/testKLU_RocSolver_FGMRES.cpp:144]:
```suggestion
  std::cout << "Results (first matrix): " << std::scientific <<std::endl<<std::endl;
```

File: tests/functionality/testKLU_RocSolver_FGMRES.cpp [hunk 9]
```diff
@@ -141,7 +141,7 @@ int main(int argc, char *argv[])
...
   std::cout<<"\t ||b-A*x||_2                 : " << std::setprecision(16) << normRmatrix1    << " (residual norm)" << std::endl;
```
Comment (kswirydo) [inline tests/functionality/testKLU_RocSolver_FGMRES.cpp:145]:
```suggestion
  std::cout << "\t ||b-A*x||_2                 : " << std::setprecision(16) << normRmatrix1    << " (residual norm)" << std::endl;
```

File: tests/functionality/testKLU_RocSolver_FGMRES.cpp [hunk 10]
```diff
@@ -141,7 +141,7 @@ int main(int argc, char *argv[])
...
   std::cout<<"\t ||b-A*x||_2  (CPU)          : " << std::setprecision(16) << normRmatrix1CPU << " (residual norm)" << std::endl;
```
Comment (kswirydo) [inline tests/functionality/testKLU_RocSolver_FGMRES.cpp:146]:
```suggestion
  std::cout << "\t ||b-A*x||_2  (CPU)          : " << std::setprecision(16) << normRmatrix1CPU << " (residual norm)" << std::endl;
```

File: tests/functionality/testKLU_RocSolver_FGMRES.cpp [hunk 11]
```diff
@@ -141,7 +141,7 @@ int main(int argc, char *argv[])
...
   std::cout<<"\t ||b-A*x||_2/||b||_2         : " << normRmatrix1/normB1   << " (scaled residual norm)"             << std::endl;
```
Comment (kswirydo) [inline tests/functionality/testKLU_RocSolver_FGMRES.cpp:147]:
```suggestion
  std::cout << "\t ||b-A*x||_2/||b||_2         : " << normRmatrix1/normB1   << " (scaled residual norm)"             << std::endl;
```

**Review by kswirydo — COMMENTED** [review]
See comments.

File: resolve/matrix/Coo.cpp [hunk 12]
```diff
@@ -84,27 +84,27 @@ namespace ReSolve
 
     if (memspaceOut == memory::HOST) {
       //check if cpu data allocated	
-      if (h_row_data_ == nullptr) {
+      if ((h_row_data_ == nullptr) && (h_col_data_ == nullptr)) {
```
[the remark this answers is not in the document]
Comment (pelesh) [reply resolve/matrix/Coo.cpp:87]: If we check them one by one, where we should set the "owns data" flag? This way, we allocate them both and set the right flag in the same scope. If it is a pathological case where only one pointer is null, the next command (copy data) will pick it up and report the error. That was my thinking here.

File: resolve/SystemSolver.cpp [hunk 13]
```diff
@@ -411,42 +411,61 @@ namespace ReSolve
     assert(rhs->getSize() == resVector_->getSize());
     real_type norm_b  = 0.0;
     real_type resnorm = 0.0;
-
+    memory::MemorySpace ms = memory::HOST;
```
[the remark this answers is not in the document]
Comment (pelesh) [reply resolve/SystemSolver.cpp:414]: Yes, in lines 434-436.

File: resolve/matrix/Coo.cpp [hunk 14]
```diff
@@ -114,32 +114,24 @@ namespace ReSolve
         mem_.copyArrayHostToHost(h_col_data_, col_data, nnz_current);
         mem_.copyArrayHostToHost(h_val_data_, val_data, nnz_current);
         h_data_updated_ = true;
-        owns_cpu_data_ = true;
```
[the remark this answers is not in the document]
Comment (pelesh) [reply resolve/matrix/Coo.cpp:117]: My understanding is you own it when you allocate it. The "data updated" flag will indicate you set it to correct values.

File: resolve/matrix/Coo.cpp [hunk 15]
```diff
@@ -84,27 +84,27 @@ namespace ReSolve
...
       if (h_val_data_ == nullptr) {
         this->h_val_data_ = new real_type[nnz_current];
+        owns_cpu_vals_ = true;
       }
     }
 
     if (memspaceOut == memory::DEVICE) {
       //check if cuda data allocated
-      if (d_row_data_ == nullptr) {
+      if ((d_row_data_ == nullptr) && (d_col_data_ == nullptr)) {
```
[the remark this answers is not in the document]
Comment (pelesh) [reply resolve/matrix/Coo.cpp:100]: See comment on line 87.

File: resolve/matrix/Csc.cpp [hunk 16]
```diff
@@ -80,27 +80,27 @@ namespace ReSolve
 
     if (memspaceOut == memory::HOST) {
       //check if cpu data allocated
-      if (h_col_data_ == nullptr) {
+      if ((h_col_data_ == nullptr) && (h_row_data_ == nullptr)) {
```
[the remark this answers is not in the document]
Comment (pelesh) [reply resolve/matrix/Csc.cpp:83]: See comment for `Coo.cpp`, line 87.

File: resolve/matrix/Sparse.cpp [hunk 17]
```diff
@@ -216,36 +216,34 @@ namespace ReSolve { namespace matrix {
...
 
     switch(control)  {
       case 0: //cpu->cpu
         mem_.copyArrayHostToHost(h_val_data_, new_vals, nnz_current);
         h_data_updated_ = true;
-        owns_cpu_vals_ = true;
         break;
       case 2://cuda->cpu
         mem_.copyArrayDeviceToHost(h_val_data_, new_vals, nnz_current);
         h_data_updated_ = true;
```
[the remark this answers is not in the document]
Comment (pelesh) [reply resolve/matrix/Sparse.cpp:238]: I suggest we handle this in a separate PR.

File: resolve/matrix/Csr.cpp [hunk 18]
```diff
@@ -97,27 +97,33 @@ namespace ReSolve
 
     if (memspaceOut == memory::HOST) {
       //check if cpu data allocated
-      if (h_row_data_ == nullptr) {
-        this->h_row_data_ = new index_type[n_ + 1];
+      if ((h_row_data_ == nullptr) != (h_col_data_ == nullptr)) {
```
[the remark this answers is not in the document]
Comment (pelesh) [reply resolve/matrix/Csr.cpp:100]: I wasn't sure if it is an overkill. Should we use it everywhere or remove it from Csr?

[the remark this answers is not in the document]
Comment (kswirydo) [reply resolve/matrix/Csr.cpp:100]: use it everywhere I think (?)

[the remark this answers is not in the document]
Comment (pelesh) [reply resolve/matrix/Csr.cpp:100]: Addressed in [7fa6682](https://github.com/ORNL/ReSolve/pull/124/commits/7fa6682c9a0f3f99b89705f62f26e8c4fcce2c3b).

File: resolve/matrix/Csr.cpp [hunk 19]
```diff
@@ -97,27 +97,33 @@ namespace ReSolve
...
         this->h_val_data_ = new real_type[nnz_current];
+        owns_cpu_vals_ = true;
       }
     }
 
     if (memspaceOut == memory::DEVICE) {
       //check if cuda data allocated
-      if (d_row_data_ == nullptr) {
-        mem_.allocateArrayOnDevice(&d_row_data_, n_ + 1); 
+      if ((d_row_data_ == nullptr) != (d_col_data_ == nullptr)) {
```
[the remark this answers is not in the document]
Comment (pelesh) [reply resolve/matrix/Csr.cpp:116]: Addressed in [7fa6682](https://github.com/ORNL/ReSolve/pull/124/commits/7fa6682c9a0f3f99b89705f62f26e8c4fcce2c3b).

File: resolve/matrix/Csr.cpp [hunk 20]
```diff
@@ -212,40 +210,42 @@ namespace ReSolve
       case HOST:
         //check if we need to copy or not
         if ((d_data_updated_ == true) && (h_data_updated_ == false)) {
-          if (h_row_data_ == nullptr) {
-            h_row_data_ = new index_type[n_ + 1];      
+          if ((h_row_data_ == nullptr) != (h_col_data_ == nullptr)) {
```
[the remark this answers is not in the document]
Comment (pelesh) [reply resolve/matrix/Csr.cpp:213]: Addressed in [7fa6682](https://github.com/ORNL/ReSolve/pull/124/commits/7fa6682c9a0f3f99b89705f62f26e8c4fcce2c3b).

File: resolve/matrix/Csr.cpp [hunk 21]
```diff
@@ -212,40 +210,42 @@ namespace ReSolve
...
           h_data_updated_ = true;
-          owns_cpu_data_ = true;
-          owns_cpu_vals_ = true;
         }
         return 0;
       case DEVICE:
         if ((d_data_updated_ == false) && (h_data_updated_ == true)) {
-          if (d_row_data_ == nullptr) {
-            mem_.allocateArrayOnDevice(&d_row_data_, n_ + 1); 
+          if ((d_row_data_ == nullptr) != (d_col_data_ == nullptr)) {
```
[the remark this answers is not in the document]
Comment (pelesh) [reply resolve/matrix/Csr.cpp:233]: Addressed in [7fa6682](https://github.com/ORNL/ReSolve/pull/124/commits/7fa6682c9a0f3f99b89705f62f26e8c4fcce2c3b).

File: resolve/matrix/Sparse.cpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (pelesh) [reply resolve/matrix/Sparse.cpp:238]: See #126

**Pull request merged**